### ISSUE TITLE
shiny::devmode and logging the R console to a log file via sink() - traces absent

### ISSUE THREAD

**Issue by nir4most** [rstudio/shiny#3677]
I would like to understand a best practice way to log a shiny app, so as to track the console, which will also include shiny error traces from shiny::devmode. If this is not currently possible, I would request this as a feature.

Here is a reprex to illustrate the issue:
```
library(shiny)
shiny::devmode()
ui <- fluidPage(
  actionButton("simple", "Just to add something normal to the log"),
  actionButton("force_error_button", "Force an Error")
)
close_sink_and_quit <- function(){
  sink(file = NULL,type = "message") 
  sink(file = NULL,type = "output") 
  close(con = flog)
  stopApp()
}

options(error = close_sink_and_quit)
options(shiny.error = close_sink_and_quit)


flog <- file("my_shiny_log", open = "wt")
sink(file = flog, split = TRUE)
sink(file = flog, type = "message")

server <- function(input, output, session) {

  onStop(function() {cat("Session stopped\n") ;close_sink_and_quit()})
  

  observeEvent(input$simple,
               cat("\npressed for",input$simple))

  
  observeEvent(input$force_error_button, {
    stop(paste("Error Forced at ", Sys.time()))
  })
}

shinyApp(ui, server)
```
if I refrain from pressing my force error button, but hit the simple button a few times,
I happily get the corresponding message in my console, as well as in "my_shiny_log" file .

if I force an error, although I see a useful stack trace in the console, this is _not_ represented in the "my_shiny_log" file. 
I've been pulling my hair out on this point :D

example of trace that shows in console but not in the sink log file : 
```
Warning: Error in <observer:observeEvent(input$force_error_button)>: Error Forced at  2022-07-26 12:43:12
  76: h
  75: .handleSimpleError
  74: stop
  73: <observer:observeEvent(input$force_error_button)> [#11]
  72: valueFunc
  71: ..stacktraceon..
  70: contextFunc
```

so to summarise, is it possible to duplicate the full console *including devmode's stacktrace* in a `sink`ed logging file?, or would this require a new devmode feature/bugfix to achieve ?

I had it suggested to me that a combination of wrapping the app inside a callr/processx might be a way to go, though this seemed rather heavyweight to me... thoughts ?

**Comment by nirguk** [comment]
Its been more than a year of living without this; can anyone possibly help ?

**Comment by gadenbuie** [comment]
> if I force an error, although I see a useful stack trace in the console, this is _not_ represented in the "my_shiny_log" file.

I think this is likely because the `sink()` is added in the wrong context, i.e. inside the app code rather than entirely outside of the code that's actually running the app.

Here are two approaches you could take:

First, you could create a second `run-app.R` file with the code that runs your app. Move all of the sink logic into that second file.

```r
# app.R
library(shiny)

ui <- fluidPage(
  actionButton("simple", "Just to add something normal to the log"),
  actionButton("force_error_button", "Force an Error")
)

server <- function(input, output, session) {
  observeEvent(
    input$simple,
    cat("\npressed for", input$simple)
  )


  observeEvent(input$force_error_button, {
    stop(paste("Error Forced at ", Sys.time()))
  })
}

shinyApp(ui, server)
```

```r
# run-app.R
run_app <- function() {
  flog <- file("my_shiny.log", open = "wt")
  on.exit({close(flog)})
  sink(file = flog, split = TRUE)
  sink(file = flog, type = "message")

  shiny::runApp("app.R")
}

shiny::devmode()
run_app()
```

Then run the app from the Terminal with the following to collect all logs into a single file.

```bash
Rscript run-app.R
```

Second, you could use `RScript` to run the app but use Bash output redirects to store pipe the output into log files:

```bash
Rscript -e 'shiny::runApp("app.R")' > shiny.log 2>&1
```

The `> shiny.log` creates a writes the log messages into `shiny.log` (deleting previous log contents, if you want to keep those use `>> shiny.log`). The `2>&1` redirects the error message stream (`2`) to the standard output stream (`&1`).

**Comment by nirguk** [comment]
I'm afraid I dont follow this : I think this is likely because the sink() is added in the wrong context, i.e. inside the app code rather than entirely outside of the code that's actually running the app.

If I do a superficial restructure of my original example.  moving the sink code to the top , and wrapping the rest in a function call, and then running that function; the behaviour of the system is exactly as in my original setup. 

I will reproduce that example here : 
```
```

moreover, when I take your proposal and run it from Rstudio, the app launches and the behaviour is somehow better in that the log file is more complete; yet the output is no longer shown in the console (which for the purposes of being a developer on my project is a regression, as it means to tackle an error, I need to go additional step and pull up a plain text log, rather than immediately glance at my color coded console) 

furthermore, when I launch it via the Rscript.exe concept,  I simply dont get an app launching at all.  but perhaps I'm overlooking something on that... its not my priority at this time

**Comment by ismirsehregal** [comment]
Regarding your second concern you could use the tee utility to duplicates the data streams and redirect them to the console and the log file:

`Rscript -e 'shiny::runApp("app.R")' 2>&1 | tee shiny.log`

In the end the above approaches are quite similar to my [initial answer](https://forum.posit.co/t/shiny-devmode-and-writing-logs-sink/143106/2).

**Comment by nirguk** [comment]
I suppose I may just have a bee in my bonnet , but it seems to me that shiny should support all outputs going both console and a nominated logfile in a direct/straightforward manner, just an option. , rather than require additional wrapping, and certainly not need to be wrapped at the Operating System level....(Rscript etc) but perhaps I am simply asking for too much or out of step with the priorities of others in the shiny world. 

Is this fundamentally a Base R issue; going back to the nature of sink; and the fact that its documented to only support split on type 'output' and not on type 'message' ?

**Comment by ismirsehregal** [comment]
Just a side note: are you aware of shiny's new `onUnhandledError` callback?

**Comment by nirguk** [comment]
Thank you, I had not, I will look into that... ; appreciate you.
